src/passes/Heap2Local.cpp

```diff
@@ -783,24 +785,21 @@ struct Struct2Local : PostWalker<Struct2Local> {
       return;
     }
 
-    // It is safe to optimize out this RefCast, since we proved it
-    // contains our allocation and we have checked that the type of
-    // the allocation is a subtype of the type of the cast, and so
-    // cannot trap.
-    replaceCurrent(curr->ref);
+    // We know this RefCast receives our allocation, so we can see whether it
+    // succeeds or fails.
+    if (Type::isSubType(allocation->type, curr->type)) {
+      // The cast succeeds, so it is a no-op, and we can skip it.
+      replaceCurrent(curr->ref);
```

Comment:
~Even if the cast is known to succeed, we might still need it to make the static types work out, right? This replacement is only valid if `curr->ref->type` is a subtype of `curr->type`, not necessarily if `allocation->type` is a subtype.~

~This isn't a new issue, so I must be missing something that makes this safe.~

Aha, it's because we're going to remove the allocation entirely. This is worth a comment!

Reply:
Sounds good, I added a comment.